src/basis/Xdg.ml

```diff
@@ -0,0 +1,65 @@
+open File
+
+(** The scheme refers to how the default directory should be determined
+    when XDG_CONFIG_HOME or XDG_CACHE_HOME is missing.
+
+    It does not correspond to the actual OS that is running. For example, the
+    [Linux] scheme covers all BSD-like systems and Cygwin on Windows. *)
+type scheme = MacOS | Linux | Windows
+
+let uname_s =
+  lazy begin
+    try
+      let ic = Unix.open_process_args_in "uname" [|"-s"|] in
+      try
+        let res = String.trim @@ input_line ic in
+        close_in_noerr ic;
+        Some res
+      with
+      | _ ->
+        close_in_noerr ic;
+        None
+    with
+    | _ ->
+      None
+  end
+
+let guess_scheme =
+  lazy begin
+    match Sys.os_type with
+    | "Unix" ->
+      begin
+        match Lazy.force uname_s with
+        | Some "Darwin" -> MacOS
+        | _ -> Linux
+      end
+    | "Cygwin" -> Linux
+    | "Win32" -> Windows
+    | _ -> Linux
+  end
+
+(* XXX I did not test the following code on different platforms. *)
+let get_config_home ~app_name =
+  match Sys.getenv_opt "XDG_CONFIG_HOME" with
+  | Some dir -> dir/app_name
+  | None ->
+    match Lazy.force guess_scheme with
+    | Linux ->
+      Sys.getenv "HOME"/".config"/app_name
+    | MacOS ->
```

Comment:
btw... developers who use macos generally prefer to put their configuration files in the linux style, even though the application support is used by most applications. that’s because most of our tools are coming from linux at first. 

but i guess it’s fine to support this scheme.

Reply:
Interesting... let me think about it.

On the other hand, many Linux tools seem to be migrating to `XDG_CONFIG_HOME`, or at least as an option. If so, maybe setting `XDG_CONFIG_HOME` will unify everything again?

Reply:
Let me merge this first, but I still want to learn more.